src/export/markup.rs

```diff
@@ -0,0 +1,77 @@
+use crate::benchmark::relative_speed::BenchmarkResultWithRelativeSpeed;
+use crate::output::format::format_duration_value;
+use crate::util::units::Unit;
+
+pub trait MarkupFormatter {
+    fn table(&self, data: &[&[&str]]) -> String {
+        let head: &[&str] = data.first().unwrap();
+        let tail = &data[1..];
+
+        // emit header
+        let table = &mut self.table_data(head);
+
+        // emit horizontal line
+        table.push_str(&self.table_line(head.len()));
+
+        // emit data rows
+        for row in tail {
+            table.push_str(&self.table_data(row))
+        }
+
+        table.to_string()
+    }
+
+    fn table_data(&self, data: &[&str]) -> String;
+
+    fn table_line(&self, size: usize) -> String;
+
+    fn command(&self, size: &str) -> String;
+
+    fn results(
+        &self,
+        entries: &[BenchmarkResultWithRelativeSpeed],
+        unit: Unit,
+    ) -> Vec<Vec<String>> {
```

Comment:
I don't really understand why we need this function (in this form). Why do we need the intermediate `Vec<Vec<String>>`? Couldn't this function directly call `self.table_data(…)` etc.?

Reply:
Yes, it would be possible, but the overall idea here is to introduce a separation of concerns: 
1) derive/prepare the values (in this case a `Vec<Vec<String>>`) which should be used for exporting;
2) use the values and export them in the desired format.

The advantage of this implementation would have two impacts for upcoming changes:
1) we can provide different kind of `results` creation functionalities without interfering with the actual exporting (writing of tables);
2) we can optimize the exports to derive a result and pass this prepared values just into the different exporters if there are multiple selected.

Reply:
> Yes, it would be possible, but the overall idea here is to introduce a separation of concerns:
> 
>     1. derive/prepare the values (in this case a `Vec<Vec<String>>`) which should be used for exporting;

Ok, but then why does it need to be part of this trait? If we want actual separation of concerns, I would only move the markup-format-specific functionality into that trait. Another part of the code could then operate on a generic `dyn MarkupFormatter` and call the different `table*` commands as appropriate.

> The advantage of this implementation would have two impacts for upcoming changes:
> 
>     1. we can provide different kind of `results` creation functionalities without interfering with the actual exporting (writing of tables);

Not sure I understand...?

> we can optimize the exports to derive a result and pass this prepared values just into the different exporters if there are multiple selected.

Ok, agreed. It seems unnecessary to do it multiple times, but performance is really not a big concern here.

Reply:
> Ok, but then why does it need to be part of this trait? If we want actual separation of concerns, I would only move the markup-format-specific functionality into that trait. Another part of the code could then operate on a generic dyn MarkupFormatter and call the different table* commands as appropriate.

It is currently part of this trait, which we can (maybe) refactor out later in another step into another trait.
Furthermore, it IMHO makes the reading of the actual exporting functionality easier, because the separation I'm referring to is that the creation of results has a dedicated call here:
https://github.com/sharkdp/hyperfine/blob/0193cdeedaba28435402fceb0adec81e1042cdf5/src/export/markdown.rs#L40

And then some lines later the actual writing of that "formatted" data.
https://github.com/sharkdp/hyperfine/blob/0193cdeedaba28435402fceb0adec81e1042cdf5/src/export/markdown.rs#L46

> Not sure I understand...?

Hereby I'm referring to the upcoming PRs I've already prepared, to provide and support not only "relative" based summary results to export. The next PRs include another results generation called "runs" to create the appropriate table data for complete benchmark runs etc.